### PR TITLE
Implementation of "constant color output" pass for PIX

### PR DESCRIPTION
**Pull request by jeffnn**
Adds a pass that seeks the "return" statement in the entry-point function, then adds additional writes of a literal constant to the output (color) register. No effort is made to remove any existing writes to the output color register.

The test validates that the original and new writes exist. (Note that no specific values in the new outputs are checked for: an imminent modification of this pass will parameterize this color, and the test will be update to match at that time.)

### CODE REVIEW

File: lib/HLSL/DxilOutputColorBecomesConstant.cpp [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+    float color[4] = { r, g, b, a };
+
+    DxilModule &DM = M.GetOrCreateDxilModule();
+
+    auto pEntrypoint = DM.GetEntryFunction();
+
+    BasicBlock * pBasicBlock = pEntrypoint->begin();
+
+    for (auto pInstruction = pBasicBlock->begin(); pInstruction != pBasicBlock->end(); pInstruction++) {
```
Comment (marcelolr) [inline lib/HLSL/DxilOutputColorBecomesConstant.cpp]: You can walk the uses of the store output function.

File: lib/HLSL/DxilOutputColorBecomesConstant.cpp [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+        IRBuilder<> Builder(pInstruction);
+
+        for (unsigned i = 0; i < _countof(color); ++i) {
+
+          // Create a new hlsl operation as a factory for creating a reference to the store-output "global function" and the required float constants
+          OP *hlslOP = DM.GetOP();
+
+          Function *pOutputFunction = hlslOP->GetOpFunc(DXIL::OpCode::StoreOutput, Builder.getFloatTy());
+
+          std::unique_ptr<OP> hlslOp = std::make_unique<OP>(M.getContext(), &M);
```
Comment (marcelolr) [inline lib/HLSL/DxilOutputColorBecomesConstant.cpp]: You can simply reuse the original OP helper object.

File: lib/HLSL/DxilOutputColorBecomesConstant.cpp [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+          Constant *OpArg = hlslOp->GetU32Const((unsigned)DXIL::OpCode::StoreOutput);
+
+          // Prepare the argument array, of which the first two elements remain constant for each of the four pixel-elements we're going to set
+          SmallVector<Value *, 5> Args;
+          Args.push_back(OpArg);
+          Args.push_back(Builder.getInt32(0)); //todo: one of these is probably an index into the output signature... so we'll have to look up the proper index
```
Comment (marcelolr) [inline lib/HLSL/DxilOutputColorBecomesConstant.cpp]: From the DxilModule, the output signature will have elements with Color - those are the ones you want. The column count and offset will also allow you to map the components you're interested in.

File: lib/HLSL/DxilOutputColorBecomesConstant.cpp [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    if (!targetFound)
+    {
+      return false;
+    }
+
+    auto pEntrypoint = DM.GetEntryFunction();
+
+    BasicBlock * pBasicBlock = pEntrypoint->begin();
+
+    IRBuilder<> Builder(pBasicBlock->begin());
```
Comment (jeffnn) [inline lib/HLSL/DxilOutputColorBecomesConstant.cpp]: All I'm using this builder for is to get an instance of the float type. Is there a better way?

File: lib/HLSL/DxilOutputColorBecomesConstant.cpp [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+          APInt outputRow = pOutputRow->getValue();
+          outputRow;
+
+          Value * pOutputColumnOperand = instruction->getOperand(hlsl::DXIL::OperandIndex::kStoreOutputColOpIdx);
+          ConstantInt * pOutputColumn = cast<ConstantInt>(pOutputColumnOperand);
+          APInt outputColumn = pOutputColumn->getValue();
+
+          Constant * pFloatConstant = hlslOP->GetFloatConst(color[*outputColumn.getRawData()]);
+
+          instruction->setOperand(hlsl::DXIL::OperandIndex::kStoreOutputValOpIdx, pFloatConstant);
```
Comment (jeffnn) [inline lib/HLSL/DxilOutputColorBecomesConstant.cpp]: I'm not sure how to integrate the row and column offsets obtained from the output signature. It seems like the columns will always be 0,1,2,3 anyway?

File: tools/clang/test/HLSL/pix/constantcolor.hlsl [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+
+// Added override output color:
+// CHECK: call void @dx.op.storeOutput.f32(i32 5, i32 0, i32 0, i8 0, float 0
+// CHECK: call void @dx.op.storeOutput.f32(i32 5, i32 0, i32 0, i8 1, float 0
+// CHECK: call void @dx.op.storeOutput.f32(i32 5, i32 0, i32 0, i8 2, float 0
+// CHECK: call void @dx.op.storeOutput.f32(i32 5, i32 0, i32 0, i8 3, float 1
+
+[RootSignature("")]
+float4 main() : SV_Target {
+    return float4(0,0,0,0);
```
Comment (jeffnn) [inline tools/clang/test/HLSL/pix/constantcolor.hlsl]: I'll extend this test as soon as I've stabilized the pass itself (i.e. figured out how to elicit non-zero starting rows etc.)

File: lib/HLSL/DxilOutputColorBecomesConstant.cpp (hunk 4 above)
Comment (tex3d) [inline lib/HLSL/DxilOutputColorBecomesConstant.cpp]: LLVMContext Ctx = M.getContext();
... Type::getFloatTy(Ctx)

File: lib/HLSL/DxilOutputColorBecomesConstant.cpp (hunk 5 above)
Comment (tex3d) [inline lib/HLSL/DxilOutputColorBecomesConstant.cpp]: If you want an absolute index into the packed signature, you'll have to look up the packing location of the element being stored to.

Note: for convenience, you can use DxilInst_StoreOutput (see DxilInstructions.h) to wrap instruction and provide more convenient accessors.

To do this, do something like:
DxilInst_StoreOutput SO(instruction);

DxilSignatureElement &E = DM->GetOutputSignature().GetElement(SO.get_outputtSigId());
    (note: spelling error in outputtSigId - we should correct that in hctdb.py!)

The packing location is here: E.GetStartRow(), E.GetStartCol()

Add these to the row, col offset into the element to get the absolute location in the packed signature.

File: lib/HLSL/DxilOutputColorBecomesConstant.cpp [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+    {
+      Blue = atof(option.second.data());
+    }
+    else if (0 == option.first.compare("constant-alpha"))
+    {
+      Alpha = atof(option.second.data());
+    }
+    else if (0 == option.first.compare("mod-mode"))
+    {
+      Mode = static_cast<VisualizerInstrumentationMode>(atoi(option.second.data()));
```
Comment (tex3d) [inline lib/HLSL/DxilOutputColorBecomesConstant.cpp]: I don't see where this is used.  Is this for future use?

Comment (jeffnn) [inline lib/HLSL/DxilOutputColorBecomesConstant.cpp]: Yes. I wanted to make one change on the PIX side that passed these params in and be able to test it. My next change will be to implement the mod modes.

**Pull request merged**